drivers/serial/uart_ipc.c

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/ipc/ipc_service.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#define DT_DRV_COMPAT nordic_nrf_ipc_uart
+
+#define BIND_TIMEOUT CONFIG_IPC_UART_BIND_TIMEOUT_MS
+
+LOG_MODULE_REGISTER(ipc_uart, CONFIG_IPC_UART_LOG_LEVEL);
+
+struct uart_ipc_config {
+	const struct device *ipc;
+	const char *ept_name;
+};
+
+struct uart_ipc_data {
+	struct ipc_ept ept;
+	struct ipc_ept_cfg ept_cfg;
+	struct k_event ept_bond;
+	struct k_work ipc_send_work;
+	struct ring_buf *tx_ringbuf;
+	struct ring_buf *rx_ringbuf;
+};
+
+static const struct uart_ipc_config *get_config(const struct device *dev)
+{
+	return dev->config;
+}
+
+static struct uart_ipc_data *get_data(const struct device *dev)
+{
+	return dev->data;
+};
+
+static void ipc_send_work_handler(struct k_work *work)
+{
+	int ret;
+	uint32_t data_size;
+	uint8_t *data;
+	struct uart_ipc_data *dev_data;
+
+	dev_data = CONTAINER_OF(work, struct uart_ipc_data, ipc_send_work);
+
+	data_size = ring_buf_get_claim(dev_data->tx_ringbuf, &data,
+				       ring_buf_size_get(dev_data->tx_ringbuf));
+	if (data_size == 0) {
+		/* Nothing to send. */
+		return;
+	}
+
+	ret = ipc_service_send(&dev_data->ept, data, data_size);
+	if (ret == -ENOMEM) {
+		k_work_submit(&dev_data->ipc_send_work);
```

Comment:
```suggestion
                k_msleep(1);
		k_work_submit(&dev_data->ipc_send_work);
```
It seems to me that it is worth waiting before the next attempt to send.

Reply:
It is not. The System Workqueue thread has high priority and waiting here delays almost whole system. I understand why you suggest to add delay here but it must be done by using delayed workqueue

Reply:
Done

Reply:
Thanks. Actually, reschedule with a delay is a better approach.